pulsar-broker/src/test/java/org/apache/pulsar/client/api/TopicReaderTest.java

```diff
@@ -510,4 +513,114 @@ public void testReaderNonDurableIsAbleToSeekRelativeTime() throws Exception {
         reader.close();
         producer.close();
     }
+
+    @Test
+    public void testReaderIsAbleToSeekWithTimeOnBeginningOfTopic() throws Exception {
+        final String topicName = "persistent://my-property/my-ns/ReaderSeekWithTimeOnBeginningOfTopic";
+        final int numOfMessage = 10;
+
+        Producer<byte[]> producer = pulsarClient.newProducer()
+                .topic(topicName).create();
+
+        for (int i = 0; i < numOfMessage; i++) {
+            producer.send(String.format("msg num %d", i).getBytes());
+        }
+
+        Reader<byte[]> reader = pulsarClient.newReader().topic(topicName)
+                .startMessageId(MessageId.earliest).create();
+
+        assertTrue(reader.hasMessageAvailable());
+
+        // Read all messages the first time
+        for (int i = 0; i < numOfMessage; i++) {
+            Message<byte[]> message = reader.readNext();
+            Assert.assertEquals(message.getData(), String.format("msg num %d", i).getBytes());
+        }
+
+        assertFalse(reader.hasMessageAvailable());
+
+        // Perform cursor reset by time
+        reader.seek(RelativeTimeUtil.parseRelativeTimeInSeconds("-1m"));
+
+        // FIXME: This sleep is necessary because future completed by seekAsync() might complete with a state different
+        //        than Ready, this is from the fact that consumer reset makes a disconnect from broker in order to be
+        //        able to do the reset().
+        Thread.sleep(1000);
```

Comment:
Didn't understand what you mean by - 
`this is from the fact that consumer reset makes a disconnect from broker in order to be able to do the reset()`

Reply:
If you are waiting for a reconnect to complete then I would suggest wait for upto 10 seconds with 20 retries (Each of 500 ms)

```
for (int i = 0; i<20 && !reader.isConnected(); i++) {
    Thread.sleep(500);
}

```

Reply:
Agree with retries until connection is ready, changing it towards that. :+1: 

> Didn't understand what you mean by `this is from the fact that consumer reset makes a disconnect from broker in order to be able to do the reset()`

Sorry, that explanation was a little poor.

What I'm trying to explain here is:
 
A `reset()` would result on a consumer disconnect on any subscription (a reconnect effectively) [see here](https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java#L393).
Because of that, the subscription which made the `reset()` might (depends on unlucky timing) result on a finished `seek()` from [here](https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java#L1380), but at that very moment, the consumer might be on a `Connecting` state due to reconnect, because of that, I added the `sleep()` in order to wait for such reconnect, in a future we might need to add the ability to complete that future waiting for consumer state being `Ready`.

please let me know if I made not this clear enough again

Reply:
I realized that we can remove `sleep()` and assertions testing if the consumers is connected and the number of messages in queue, because if we call `message.receive()` after `seek()` will wait until a message has arrive without no matter if the consumer was not connected, and is an implicit way to check the number of messages received without asserting the incoming message queue size.

So, we can leave the `sleep()` and do these asserts, or we can remove `sleep()` and just receive the messages we expect to receive, what path do you like to take?

Reply:
I took the last path since it makes things easier to understand, please let me know if you think differently